src/probe_identity.rs

```diff
@@ -0,0 +1,186 @@
+    Probe {name: String},
+    None
+}
+
+fn probe_from_bpm(input: &str) -> Result<String>
+{
+    let opening_paren = input.find('(');
+    let closing_paren = input.find(')');
+
+    match (opening_paren, closing_paren) {
+        (None, None) => Ok(NATIVE_BPM.into()),
+        (Some(opening_paren), Some(closing_paren)) => {
+            if opening_paren > closing_paren {
+                Err(eyre!("'(' is defined after ')'"))
+            } else {
+               Ok(input[opening_paren+1..closing_paren].to_string().to_lowercase())
+            }
+        }
+        (Some(_), None) => Err(eyre!("No ')' defined")),
+        (None, Some(_)) => Err(eyre!("No '(' defined")),
+    }
+}
+
+fn version_from_bpm(input: &str) -> Result<String>
+{
+    let start_index = input.rfind(' ')
+        .ok_or_else(|| eyre!("Invalid version string"))?;
+
+    let version = input[start_index + 1..].to_string();
+
+    if version.trim().is_empty() {
+        return Err(eyre!("Version is empty or only whitespace"));
+    }
+
+    Ok(version)
+}
+
+impl From<String> for ProbeIdentity
+{
+    fn from(identity: String) -> Self
+    {
+        if identity == BMP_PRODUCT_STRING {
+            ProbeIdentity::Probe {
+                name: NATIVE_BPM.into()
+            }
+        } else if identity.starts_with(BMP_PRODUCT_STRING) {
+            let probe_result = probe_from_bpm(&identity[BMP_PRODUCT_STRING_LENGTH..]);
+
+            let probe = match probe_result {
+                Ok(probe) => probe,
+                Err(error) => {
+                    warn!("Error while parsing probe string: {}", error);
+                    return  ProbeIdentity::None
+                }
+            };
+            
+            let version = version_from_bpm(&identity[BMP_PRODUCT_STRING_LENGTH..]);
+
+            match version {
+                Ok(version) => ProbeIdentity::ProbeVersion {
+                    name: probe,
+                    version
+                },
+                Err(error) => {
+                    warn!("Error while parsing version string: {}", error);
+                    ProbeIdentity::None
+                }
+            }            
+        } else {
+            ProbeIdentity::None
+        }
+
+    }
+}
+
+impl ProbeIdentity
+{
+    fn probe_name(&self) -> Option<String>{
+        match &self{
+            ProbeIdentity::ProbeVersion{name, ..} => Some(name.to_string()),
+            ProbeIdentity::Probe{name} => Some(name.to_string()),
+            ProbeIdentity::None => None,
+        }
+    }
+    
+    pub fn variant(&self) -> Result<Probe, Report>
+    {
+        let probe_version = self.probe_name()
+            .ok_or_else(|| eyre!("No product string discovered."))?;
+        
+        probe_version.as_str().try_into()
+    }
+
+    pub fn version(self) -> Option<String>
+    {
+        match self{
+            ProbeIdentity::ProbeVersion{version, ..} => Some(version),
+            ProbeIdentity::Probe{..} => Some("v1.6".to_string()),
```

Comment:
Note that we only know in this case that it is "pre v1.7", not what actual version it is. It is more correct to return None for that. This is why we're so careful to mark such versions "unknown" in the inspiring BMDA source: https://github.com/blackmagic-debug/blackmagic/blob/d54669408fcd3bc1c9017621d82c34ab5868bfc1/src/platforms/hosted/bmp_libusb.c#L156-L214